llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,262 @@
+  // the Preheader, so that the assumption is only valid once and not on every
+  // entry of the Loop Header.
+  IRBuilder<> Builder(dyn_cast<Instruction>(InsertionPoint));
+
+  SCEVExpander Expander(
+      SE, CurrentLoop.getLoopPreheader()->getModule()->getDataLayout(),
+      "expanded");
+
+  Value *LHS = expandValueAtMinIterCountIteration(
+      LoopCmpInstr.getOperand(0), SE, Expander, InsertionPoint, DT,
+      &CurrentLoop, MinIterCount);
+  if (!LHS)
+    return;
+  LLVM_DEBUG(dbgs() << "LHS = " << *LHS << "\n");
+
+  Value *RHS = expandValueAtMinIterCountIteration(
+      LoopCmpInstr.getOperand(1), SE, Expander, InsertionPoint, DT,
+      &CurrentLoop, MinIterCount);
+
+  if (!RHS)
+    return;
+  LLVM_DEBUG(dbgs() << "RHS = " << *RHS << "\n");
+
+  // If the false-branch-target is to the Loop Body, inverse the
+  // predicate, since the Loop Condition is inversed to remain in the Loop
+  CmpInst::Predicate Pred = LoopCmpInstr.getPredicate();
+  if (!CurrentLoop.contains(
+          dyn_cast<BranchInst>(CurrentLoop.getExitingBlock()->getTerminator())
+              ->getSuccessor(0)))
+    Pred = LoopCmpInstr.getInversePredicate();
+
+  Value *Cmp = Builder.CreateICmp(Pred, LHS, RHS);
+
+  // Insert Assumption
+  CallInst *Assumption = Builder.CreateAssumption(Cmp);
+  AC.registerAssumption(dyn_cast<AssumeInst>(Assumption));
+  LLVM_DEBUG(dbgs() << "With Comparator             :" << *Cmp << "\n"
+                    << "Assume                      :" << *Assumption << "\n");
+}
+
+/// Determine if the \param CurrentLoop is not rotated yet and Loop Iteration
+/// Count Metadata greater than 0
+std::optional<int64_t> validLoopIterCount(Loop &CurrentLoop) {
+  BasicBlock *LoopHeader = CurrentLoop.getHeader();
+
+  // Dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  } dbgs() << "LoopIterCountAssumption-Info: Function = "
+           << getFunctionAndBlockNames(*LoopHeader) << "\n");
+
+  std::optional<int64_t> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+  if (!RawMinIterationCount) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions: Loop Iteration "
+                         "Count not provided for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  const int64_t MinIterCount = *RawMinIterationCount;
+  if (MinIterCount <= 0) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Loop Iteration "
+                         "Count is smaller or equal to zero for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << getFunctionAndBlockNames(*LoopHeader) << " ("
+                    << MinIterCount << ")\n");
+  return MinIterCount - 1;
```

Comment:
You're sneakily turning an "iteration count" into an "iteration index" by subtracting `1` :) The name of the function is also a bit confusing now. Maybe name it `getValidMinIterCount()` and avoid subtracting `1`? We can do the subtractiojn right before calling `expandValueAIteration()`

Reply:
I went with the MinIterCount and subtracted it directly before calling `expandValueAIteration()`